---
title: "WholeGenome_Quality_Control"
author: "Diego Lera-Lozano and Vivian Yifan Li"
format:
  html:
    theme: darkly
    toc: true
    toc-float: true
    toc-depth: 5
    df-print: paged
    smooth-scroll: true
    citations-hover: true
editor: source  
---

# Quality Assessment

In this document we will look at assessing the quality of nanopore long reads using fastQC then looking at using chopper to trim either ends of the reads leaving us with high quality reads suitable for downstream analysis.

### FastQC

To begin we will generate a fastQC report, first we can load fastqc using Roar Collab's available modules. Then prior to generating the reports we have to concatenate all of the read files generated. Then gzip them. Then we can generate the report.

- Concatenate: ` zcat *fastq.gz > barcode01_basecalled.fastq `
- GZipping: `gzip barcode01_supbasecalled.fastq`
- FastQC: `fastqc barcode01_supbasecalled.fastq.gz`

The FastQC report has several sections but we will specifically be looking at the first four being Basic Statistics, Per base sequence quality, and Per base sequence content.

- Basic Statistics: This section will provided you with basic data that may be important downstream.
- Per base sequence quality: This section will show generated graphs assessing the quality of your reads. Here we are looking to have all of the yellow box plots above the 28 threshold. 
- Per base sequence content: In this section we are looking to make sure that the percentage of coinciding nucleotides persist until towards the end of the graph.

Examples:

-------------------------------------------------------------------

![Image 1](dorado.png){width="49.2%"} | ![Image 2](BC1_trim.png){width="49.2%"}

------------------------------------------------------------------------

On the left we






